File: dataStructure/doubleLinkedList.cpp
#include <iostream>

using namespace std;

template <typename T>
struct DNode {
  T data;
  DNode *prev;
  DNode *next;
};


template <typename T>
class DoubleLinkedList {
 private:
  DNode<T> *head;
 public:
  DoubleLinkedList() : head(nullptr) {}
  ~DoubleLinkedList() {
    DNode<T> *current = head;

    while (current != nullptr) {
      DNode<T> *next = current->next;
      delete current;
      current = next;
    }
  }

  bool isEmpty() {
    return head == nullptr;
  }

  void append(T value) {
    if (isEmpty()) {
      head = new DNode<T> {value, nullptr, nullptr};
      return;
    }

    DNode<T> *current = head;

    while (current->next != nullptr) current = current->next;
    current->next = new DNode<T> {value, current, nullptr};
  }

  void display() {
    if (isEmpty()) return;
    DNode<T> *current = head;
    while (current != nullptr) {
      cout << current->data << ' ';
      current = current->next;
    }
    cout << endl;
  }

  void displayReverse() {
    if (isEmpty()) return;

    DNode<T> *current = head;
    while (current->next != nullptr) current = current->next;
    while (current != nullptr) {
      cout << current->data << ' ';
      current = current->prev;
    }
    cout << endl;
  }

  int count() {
    int c = 0;
    DNode<T> *current = head;
    while (current != nullptr) {
      c++;
      current = current->next;
    }

    return c;
  }

  bool search(T value) {
    DNode<T> *current = head;
    while (current != nullptr) {
      if (current->data == value) return true;
      current = current->next;
    }

    return false;
  }

  void insertBefore(T item, T value) {
    if (isEmpty()) return;

    if (head->data == item) {
      DNode<T> *newDNode = new DNode<T> {value, nullptr, head};
      head = newDNode;
      head->next->prev = head;
      return;
    }

    DNode<T> *prev = head;
    DNode<T> *current = head->next;
    while (current != nullptr && current->data != item) {
      prev = current;
      current = prev->next;
    }

    if (current != nullptr) {
      DNode<T> *newDNode = new DNode<T> {value, prev, current};
      prev->next = newDNode;
      current->prev = newDNode;
    }
  }

  void deleteValue(T value) {
    if (isEmpty()) return;

    if (head->data == value) {
      DNode<T> *temp = head;
      head = head->next;
      if (head != nullptr) head->prev = nullptr;
      delete temp;
      return;
    }

    DNode<T> *prev = head;
    DNode<T> *current = head->next;
    while (current != nullptr && current->data != value) {
      prev = current;
      current = prev->next;
    }

    if (current != nullptr) {
      prev->next = current->next;
      if (current->next != nullptr) current->next->prev = prev;
      delete current;
    }
  }
};


int main() {
  DoubleLinkedList<int> arr;
  arr.append(5);
  arr.append(6);
  arr.append(7);
  arr.deleteValue(5);
  arr.append(8);
  arr.display();
  arr.displayReverse();
  return 0;
}

File: dataStructure/doubleLinkedList.java
import java.io.PrintStream;

class DNode<T> {
  T data;
  DNode<T> prev;
  DNode<T> next;

  DNode(T data, DNode<T> prev, DNode<T> next) {
    this.data = data;
    this.prev = prev;
    this.next = next;
  }
}

class DoubleLinkedList<T> {
  static final PrintStream stdout = System.out;
  static final PrintStream stderr = System.err;

  private DNode<T> head;

  public DoubleLinkedList() {
    this.head = null;
  }

  public boolean isEmpty() {
    return this.head == null;
  }

  public void append(T value) {
    if (this.isEmpty()) {
      this.head = new DNode<>(value, null, null);
      return;
    }

    DNode<T> current = this.head;
    while (current.next != null) current = current.next;
    current.next = new DNode<>(value, current, null);
  }

  public void display() {
    if (this.isEmpty()) {
      stdout.println();
      return;
    }
    DNode<T> current = this.head;
    while (current != null) {
      stdout.print(current.data + " ");
      current = current.next;
    }
    stdout.println();
  }

  public void displayReverse() {
    if (this.isEmpty()) {
      stdout.println();
      return;
    }

    DNode<T> current = this.head;
    while (current.next != null) current = current.next;
    while (current != null) {
      stdout.print(current.data + " ");
      current = current.prev;
    }
    stdout.println();
  }

  public int count() {
    int c = 0;
    DNode<T> current = this.head;
    while (current != null) {
      c++;
      current = current.next;
    }
    return c;
  }

  public boolean search(T value) {
    DNode<T> current = this.head;
    while (current != null) {
      if (current.data.equals(value)) return true;
      current = current.next;
    }
    return false;
  }

  public void insertBefore(T item, T value) {
    if (this.isEmpty()) return;

    if (this.head.data.equals(item)) {
      DNode<T> newDNode = new DNode<>(value, null, this.head);
      this.head = newDNode;
      this.head.next.prev = this.head;
      return;
    }

    DNode<T> prev = this.head;
    DNode<T> current = this.head.next;
    while (current != null && !current.data.equals(item)) {
      prev = current;
      current = prev.next;
    }

    if (current != null) {
      DNode<T> newDNode = new DNode<>(value, prev, current);
      prev.next = newDNode;
      current.prev = newDNode;
    }
    else stderr.println("Item not found!");
  }

  public void deleteValue(T value) {
    if (this.isEmpty()) return;

    if (this.head.data.equals(value)) {
      this.head = this.head.next;
      if (this.head != null) this.head.prev = null;
      return;
    }

    DNode<T> prev = this.head;
    DNode<T> current = this.head.next;
    while (current != null && !current.data.equals(value)) {
      prev = current;
      current = prev.next;
    }

    if (current != null) {
      prev.next = current.next;
      if (current.next != null) current.next.prev = prev;
    }
    else stderr.println("Value not found!");
  }
}
